### PR TITLE
Create Organization Invites

### LINKED ISSUE #28608: Allow onboarding organization members through a registration invitation link
### Description

To allow organizations to onboard members when it is not possible or desired to federate their identities from an identity provider linked to the organization, we need to allow an administrator to send a registration invitation link by e-mail so that they can use that link to automatically register to an organization as a new member.

This feature request is about:

* Given an e-mail address, send a registration invitation link via e-mail
* Register a new member to an organization once they follow the registration invitation link

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by alice-wondered**
This PR adds email invites for new and existing users to an organization. This aims to resolve #28608

### CODE REVIEW

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationMemberTest.java [hunk 1]
```diff
@@ -233,86 +232,4 @@ public void testDeleteGroupOnOrganizationRemoval() {
         assertFalse(testRealm().groups().groups().stream().anyMatch(group -> group.getName().startsWith("kc.org.")));
     }
 
-    @Test
```
Comment (pedroigor) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationMemberTest.java:236]: We should probably keep this test.

**Review by pedroigor — COMMENTED** [review]
LGTM. It would be nice to get more feedback about the contract of the REST API. 

IMO, we can handle internationalization for some messages later.

File: services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java [hunk 2]
```diff
@@ -18,6 +18,7 @@
 package org.keycloak.authentication;
 
 import org.jboss.logging.Logger;
+import org.keycloak.authentication.actiontoken.inviteorg.InviteOrgActionToken;
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:21]: Is this used?

File: services/src/main/java/org/keycloak/authentication/actiontoken/inviteorg/InviteOrgActionToken.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/authentication/actiontoken/inviteorg/InviteOrgActionToken.java:2]: Really minor - year in this header / author in class comment.

File: services/src/main/java/org/keycloak/authentication/actiontoken/inviteorg/InviteOrgActionTokenHandler.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/authentication/actiontoken/inviteorg/InviteOrgActionTokenHandler.java:2]: Same - year

File: services/src/main/java/org/keycloak/email/freemarker/FreeMarkerEmailTemplateProvider.java [hunk 5]
```diff
@@ -60,6 +60,7 @@ public class FreeMarkerEmailTemplateProvider implements EmailTemplateProvider {
     protected FreeMarkerProvider freeMarker;
     protected RealmModel realm;
     protected UserModel user;
+    protected String userEmail;
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/email/freemarker/FreeMarkerEmailTemplateProvider.java:63]: Is this used?

File: services/src/main/java/org/keycloak/organization/admin/resource/OrganizationInvitationResource.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
+package org.keycloak.organization.admin.resource;
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/organization/admin/resource/OrganizationInvitationResource.java:1]: Missing header

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationMemberTest.java (hunk 1 above)
Comment (sguilhen) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationMemberTest.java:236]: Agree, why was it going to be removed?

File: services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java [hunk 7]
```diff
@@ -101,6 +102,9 @@ public class AuthenticationProcessor {
     protected HttpRequest request;
     protected String flowId;
     protected String flowPath;
+
+
+    protected String orgToken;
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:107]: Same - is this used anywhere?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationMemberTest.java (hunk 1 above)
Comment (alice-wondered) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/organization/admin/OrganizationMemberTest.java:236]: I probably messed something up during rebase/merge and threw out the test. Will fix

File: services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java (hunk 7 above)
Comment (alice-wondered) [reply services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:107]: Will remove

File: services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java (hunk 2 above)
Comment (alice-wondered) [reply services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:21]: Will remove

File: services/src/main/java/org/keycloak/organization/admin/resource/OrganizationInvitationResource.java (hunk 6 above)
Comment (alice-wondered) [reply services/src/main/java/org/keycloak/organization/admin/resource/OrganizationInvitationResource.java:17]: Added!

**Review by pedroigor — APPROVED** [review]
Thanks, LGTM.

**Review by sguilhen — APPROVED** [review]
@alice-wondered LGTM now, thank you!

**Comment by davdenic** [comment]
Hi, is there any documentation how to use it?
thanks

**Comment by cfecherolle** [comment]
Hi, 

I found this PR when looking for a way to send email invites to newly created users, I'm also very interested in some documentation to try and use it, but I have not seen it in any release notes of versions after the merge date (2024/05/08) so I'm a bit puzzled.

**Comment by barryguvenkaya** [comment]
> Hi, is there any documentation how to use it? thanks

Looking at the file changes, ORGIVT seems to be configured as an action token. Here's the doc for action tokens: https://www.keycloak.org/docs/latest/server_development/index.html#_action_token_handler_spi

**Comment by edewit** [comment]
there is documentation here: https://www.keycloak.org/docs/latest/server_admin/index.html#inviting-users

**Pull request merged**